File: VAfilters/TPTSallenKey/python/TPTSallenKeyHPStage.py
import numpy as np


class TPTSallenKeyHPStage:
    def __init__(self, sampleRate, fc):

        self.sampleRate = sampleRate
        self.fc = fc

        wc = 2*np.pi*fc
        T = 1/sampleRate
        wa = (2/T) * np.tan(wc * T/2)

        self.g = wa * T/2
        self.G = 1 / (1 + self.g)

        self.z1 = 0

    def doFilterStage(self, xn):
        v = (xn - self.z1)
        y = v * self.G
        self.z1 = self.z1 + 2*y*self.g
        return y

    def getStorageRegisterValue(self):
        return self.z1

    def getSampleRate(self):
        return self.sampleRate

File: VAfilters/TPTSallenKey/cpp/src/TPTSallenKeyHPStage.cpp
#include "../inc/TPTSallenKeyHPStage.h"

TPTSallenKeyHPStage::TPTSallenKeyHPStage(double sampleRate, double fc) {
  this->sampleRate = sampleRate;
  this->fc = fc;

  wc = 2 * M_PI * fc;
  T = 1 / sampleRate;
  wa = (2 / T) * std::tan(wc * T / 2);

  g = wa * T / 2;
  G = 1 / (1 + g);

  z1 = 0;
}

double TPTSallenKeyHPStage::doFilterStage(double xn) {
  double v = (xn - z1);
  double y = v * G;
  z1 = z1 + 2 * y * g;
  return y;
}

double TPTSallenKeyHPStage::getStorageRegisterValue() { return z1; }

double TPTSallenKeyHPStage::getSampleRate() { return sampleRate; }